File: src/nova_audio_agent/realtime/playback.py
"""Provider-neutral renderer generation fencing and delivery acknowledgements."""

from __future__ import annotations

from collections import deque
from collections.abc import Callable
from dataclasses import dataclass
from typing import Literal

IdFactory = Callable[[], str]
DeliveryDisposition = Literal["spoken", "interrupted", "suppressed"]
MAX_PLAYBACK_FRAME_BYTES = 64 * 1024
MAX_RENDERER_TOMBSTONES = 256

_ProviderIdentity = tuple[int, str]
_RendererIdentity = tuple[str, int]


@dataclass(frozen=True, slots=True)
class PlaybackGeneration:
    session_epoch: int
    generation_epoch: int
    generation_id: str
    utterance_id: str
    response_id: str


@dataclass(frozen=True, slots=True)
class PlaybackFrame:
    utterance_id: str
    generation_epoch: int
    sequence: int
    pcm: bytes


@dataclass(frozen=True, slots=True)
class PlaybackCompletion:
    session_epoch: int
    response_id: str
    utterance_id: str
    generation_epoch: int
    text: str
    disposition: DeliveryDisposition
    started: bool = False
    played_ms: int | None = None


@dataclass(slots=True)
class _GenerationState:
    identity: PlaybackGeneration
    next_sequence: int = 0
    transcript: str = ""
    provider_terminal: bool = False
    fenced: bool = False
    delivered: bool = False
    started: bool = False
    terminal_disposition: Literal["spoken", "interrupted"] = "spoken"


class PlaybackRegistry:
    def __init__(
        self,
        *,
        id_factory: IdFactory,
        on_frame: Callable[[PlaybackFrame], None],
        on_clear: Callable[[str, int], None],
        on_alert: Callable[[str | None, int | None], None] | None = None,
    ) -> None:
        self._id_factory = id_factory
        self._on_frame = on_frame
        self._on_clear = on_clear
        self._on_alert = on_alert or (lambda _utterance_id, _generation_epoch: None)
        self._generation_epoch = 0
        self._by_provider: dict[_ProviderIdentity, _GenerationState] = {}
        self._by_renderer: dict[_RendererIdentity, _GenerationState] = {}
        self._renderer_tombstones: deque[_RendererIdentity] = deque()
        self._renderer_tombstone_set: set[_RendererIdentity] = set()
        self._current_provider_identity: _ProviderIdentity | None = None

    @property
    def current(self) -> PlaybackGeneration | None:
        if self._current_provider_identity is None:
            return None
        return self._by_provider[self._current_provider_identity].identity

    @property
    def has_unreported_fence(self) -> bool:
        """True while a fenced generation still awaits its renderer delivery report.

        The renderer may still be audibly playing that generation until its
        cleared/stopped acknowledgement arrives, so the foreground is not idle.
        """
        return any(state.fenced and not state.delivered for state in self._by_provider.values())

    def open_response(self, *, session_epoch: int, response_id: str) -> PlaybackGeneration:
        if session_epoch < 1 or not response_id:
            raise ValueError("session_epoch and response_id are required")
        provider_identity = (session_epoch, response_id)
        if provider_identity in self._by_provider:
            raise ValueError("response already has a playback generation")
        if self._current_provider_identity is not None:
            raise ValueError("another playback generation is active")
        self._generation_epoch += 1
        identity = PlaybackGeneration(
            session_epoch=session_epoch,
            generation_epoch=self._generation_epoch,
            generation_id=self._id_factory(),
            utterance_id=self._id_factory(),
            response_id=response_id,
        )
        state = _GenerationState(identity=identity)
        self._by_provider[provider_identity] = state
        self._by_renderer[self._renderer_identity(identity)] = state
        self._current_provider_identity = provider_identity
        return identity

    def push_audio(self, *, session_epoch: int, response_id: str, pcm: bytes) -> bool:
        if type(pcm) is not bytes or not pcm or len(pcm) % 2:
            raise ValueError("pcm must be non-empty aligned PCM16 bytes")
        state = self._by_provider.get((session_epoch, response_id))
        if state is None or state.provider_terminal or state.fenced or state.delivered:
            return False
        for offset in range(0, len(pcm), MAX_PLAYBACK_FRAME_BYTES):
            frame = PlaybackFrame(
                utterance_id=state.identity.utterance_id,
                generation_epoch=state.identity.generation_epoch,
                sequence=state.next_sequence,
                pcm=pcm[offset : offset + MAX_PLAYBACK_FRAME_BYTES],
            )
            self._on_frame(frame)
            state.next_sequence += 1
        return True

    def set_transcript(self, *, session_epoch: int, response_id: str, text: str) -> bool:
        state = self._by_provider.get((session_epoch, response_id))
        if state is None or state.delivered:
            return False
        state.transcript = text
        return True

    def mark_provider_terminal(
        self,
        *,
        session_epoch: int,
        response_id: str,
        disposition: Literal["spoken", "interrupted"] = "spoken",
    ) -> bool:
        state = self._by_provider.get((session_epoch, response_id))
        if state is None or state.fenced or state.delivered:
            return False
        state.provider_terminal = True
        state.terminal_disposition = disposition
        if state.next_sequence == 0:
            self._retire(state)
        return True

    def fence_current(self, *, alert: bool = False) -> PlaybackGeneration | None:
        current = self.current
        if current is None:
            if alert:
                self._on_alert(None, None)
            return None
        fenced = self.alert_fence_generation(current) if alert else self.switch_generation(current)
        return current if fenced else None

    def switch_generation(self, generation: PlaybackGeneration) -> bool:
        """Clear one exact current generation before replacement audio is emitted."""
        return self._fence_generation(generation, alert=False)

    def alert_fence_generation(self, generation: PlaybackGeneration) -> bool:
        """Alert and clear one exact current generation at the Guard deadline."""
        return self._fence_generation(generation, alert=True)

    def retire_clear_unknown(self, generation: PlaybackGeneration) -> bool:
        """Retire a fenced generation whose renderer-clear outcome is unknowable."""
        state = self._state_for_generation(generation)
        if state is None or state.delivered or not state.fenced:
            return False
        self._retire(state)
        return True

    def mark_started(self, utterance_id: str, generation_epoch: int) -> bool:
        state = self._find(utterance_id, generation_epoch)
        if state is None or state.delivered or state.started:
            return False
        state.started = True
        return True

    def record_cleared(
        self,
        utterance_id: str,
        generation_epoch: int,
        played_ms: int | None,
    ) -> PlaybackCompletion | None:
        state = self._find(utterance_id, generation_epoch)
        if state is None or state.delivered or not state.fenced:
            return None
        state.delivered = True
        audible = state.started or (played_ms is not None and played_ms > 0)
        completion = PlaybackCompletion(
            session_epoch=state.identity.session_epoch,
            response_id=state.identity.response_id,
            utterance_id=state.identity.utterance_id,
            generation_epoch=state.identity.generation_epoch,
            text=state.transcript,
            disposition="interrupted" if audible else "suppressed",
            started=state.started,
            played_ms=played_ms,
        )
        self._retire(state)
        return completion

    def ack_done(
        self,
        utterance_id: str,
        generation_epoch: int,
        played_ms: int | None = None,
    ) -> PlaybackCompletion | None:
        state = self._find(utterance_id, generation_epoch)
        if (
            state is None
            or state.fenced
            or state.delivered
            or not state.provider_terminal
            or state.next_sequence == 0
        ):
            return None
        state.delivered = True
        disposition: DeliveryDisposition = state.terminal_disposition
        if disposition == "interrupted" and not (
            state.started or (played_ms is not None and played_ms > 0)
        ):
            disposition = "suppressed"
        completion = PlaybackCompletion(
            session_epoch=state.identity.session_epoch,
            response_id=state.identity.response_id,
            utterance_id=state.identity.utterance_id,
            generation_epoch=state.identity.generation_epoch,
            text=state.transcript,
            disposition=disposition,
            started=state.started,
            played_ms=played_ms,
        )
        self._retire(state)
        return completion

    def _find(self, utterance_id: str, generation_epoch: int) -> _GenerationState | None:
        renderer_identity = (utterance_id, generation_epoch)
        if renderer_identity in self._renderer_tombstone_set:
            return None
        return self._by_renderer.get(renderer_identity)

    def _fence_generation(self, generation: PlaybackGeneration, *, alert: bool) -> bool:
        state = self._state_for_generation(generation)
        if (
            state is None
            or state.fenced
            or state.delivered
            or self._current_provider_identity != self._provider_identity(generation)
        ):
            return False
        state.fenced = True
        self._current_provider_identity = None
        if alert:
            self._on_alert(generation.utterance_id, generation.generation_epoch)
        else:
            self._on_clear(generation.utterance_id, generation.generation_epoch)
        return True

    def _state_for_generation(
        self,
        generation: PlaybackGeneration,
    ) -> _GenerationState | None:
        state = self._by_renderer.get(self._renderer_identity(generation))
        if state is None or state.identity != generation:
            return None
        return state

    def _retire(self, state: _GenerationState) -> None:
        provider_identity = self._provider_identity(state.identity)
        renderer_identity = self._renderer_identity(state.identity)
        if self._by_provider.get(provider_identity) is state:
            del self._by_provider[provider_identity]
        if self._by_renderer.get(renderer_identity) is state:
            del self._by_renderer[renderer_identity]
        if self._current_provider_identity == provider_identity:
            self._current_provider_identity = None
        self._add_renderer_tombstone(renderer_identity)

    def _add_renderer_tombstone(self, renderer_identity: _RendererIdentity) -> None:
        if renderer_identity in self._renderer_tombstone_set:
            return
        if len(self._renderer_tombstones) == MAX_RENDERER_TOMBSTONES:
            expired = self._renderer_tombstones.popleft()
            self._renderer_tombstone_set.remove(expired)
        self._renderer_tombstones.append(renderer_identity)
        self._renderer_tombstone_set.add(renderer_identity)

    @staticmethod
    def _provider_identity(generation: PlaybackGeneration) -> _ProviderIdentity:
        return generation.session_epoch, generation.response_id

    @staticmethod
    def _renderer_identity(generation: PlaybackGeneration) -> _RendererIdentity:
        return generation.utterance_id, generation.generation_epoch

File: runtime/src/playback.ts
export type DeliveryDisposition = 'spoken' | 'interrupted' | 'suppressed'

export const MAX_PLAYBACK_FRAME_BYTES = 64 * 1024
export const MAX_RENDERER_TOMBSTONES = 256

export interface PlaybackGeneration {
  readonly session_epoch: number
  readonly generation_epoch: number
  readonly generation_id: string
  readonly utterance_id: string
  readonly response_id: string
}

export interface PlaybackFrame {
  readonly utterance_id: string
  readonly generation_epoch: number
  readonly sequence: number
  readonly pcm: Uint8Array
}

export interface PlaybackCompletion {
  readonly session_epoch: number
  readonly response_id: string
  readonly utterance_id: string
  readonly generation_epoch: number
  readonly text: string
  readonly disposition: DeliveryDisposition
  readonly started: boolean
  readonly played_ms: number | null
}

interface GenerationState {
  readonly identity: PlaybackGeneration
  nextSequence: number
  transcript: string
  providerTerminal: boolean
  fenced: boolean
  delivered: boolean
  started: boolean
  terminalDisposition: Exclude<DeliveryDisposition, 'suppressed'>
}

export interface PlaybackRegistryOptions {
  readonly idFactory: () => string
  readonly onFrame: (frame: PlaybackFrame) => void
  readonly onClear: (utteranceId: string, generationEpoch: number) => void
  readonly onAlert?: (utteranceId: string | null, generationEpoch: number | null) => void
}

export class PlaybackRegistry {
  readonly #idFactory: () => string
  readonly #onFrame: (frame: PlaybackFrame) => void
  readonly #onClear: (utteranceId: string, generationEpoch: number) => void
  readonly #onAlert: (utteranceId: string | null, generationEpoch: number | null) => void
  readonly #byProvider = new Map<string, GenerationState>()
  readonly #byRenderer = new Map<string, GenerationState>()
  readonly #rendererTombstones: string[] = []
  readonly #rendererTombstoneSet = new Set<string>()
  #generationEpoch = 0
  #currentProviderIdentity: string | null = null

  constructor(options: PlaybackRegistryOptions) {
    this.#idFactory = options.idFactory
    this.#onFrame = options.onFrame
    this.#onClear = options.onClear
    this.#onAlert = options.onAlert ?? (() => undefined)
  }

  get current(): PlaybackGeneration | null {
    if (this.#currentProviderIdentity === null) return null
    return this.#byProvider.get(this.#currentProviderIdentity)?.identity ?? null
  }

  get hasUnreportedFence(): boolean {
    return [...this.#byProvider.values()].some(state => state.fenced && !state.delivered)
  }

  get rendererTombstoneCount(): number {
    return this.#rendererTombstones.length
  }

  openResponse(input: {readonly sessionEpoch: number; readonly responseId: string}): PlaybackGeneration {
    if (!Number.isInteger(input.sessionEpoch) || input.sessionEpoch < 1 || input.responseId.length === 0) {
      throw new Error('sessionEpoch and responseId are required')
    }
    const providerIdentity = providerKey(input.sessionEpoch, input.responseId)
    if (this.#byProvider.has(providerIdentity)) {
      throw new Error('response already has a playback generation')
    }
    if (this.#currentProviderIdentity !== null) {
      throw new Error('another playback generation is active')
    }
    this.#generationEpoch += 1
    const identity = Object.freeze({
      session_epoch: input.sessionEpoch,
      generation_epoch: this.#generationEpoch,
      generation_id: this.#requiredId(),
      utterance_id: this.#requiredId(),
      response_id: input.responseId,
    })
    const state: GenerationState = {
      identity,
      nextSequence: 0,
      transcript: '',
      providerTerminal: false,
      fenced: false,
      delivered: false,
      started: false,
      terminalDisposition: 'spoken',
    }
    this.#byProvider.set(providerIdentity, state)
    this.#byRenderer.set(rendererKey(identity.utterance_id, identity.generation_epoch), state)
    this.#currentProviderIdentity = providerIdentity
    return identity
  }

  pushAudio(input: {
    readonly sessionEpoch: number
    readonly responseId: string
    readonly pcm: Uint8Array
  }): boolean {
    if (!(input.pcm instanceof Uint8Array) || input.pcm.byteLength === 0 || input.pcm.byteLength % 2 !== 0) {
      throw new Error('pcm must be non-empty aligned PCM16 bytes')
    }
    const state = this.#byProvider.get(providerKey(input.sessionEpoch, input.responseId))
    if (state === undefined || state.providerTerminal || state.fenced || state.delivered) return false
    for (let offset = 0; offset < input.pcm.byteLength; offset += MAX_PLAYBACK_FRAME_BYTES) {
      this.#onFrame({
        utterance_id: state.identity.utterance_id,
        generation_epoch: state.identity.generation_epoch,
        sequence: state.nextSequence,
        pcm: input.pcm.slice(offset, offset + MAX_PLAYBACK_FRAME_BYTES),
      })
      state.nextSequence += 1
    }
    return true
  }

  setTranscript(input: {
    readonly sessionEpoch: number
    readonly responseId: string
    readonly text: string
  }): boolean {
    const state = this.#byProvider.get(providerKey(input.sessionEpoch, input.responseId))
    if (state === undefined || state.delivered) return false
    state.transcript = input.text
    return true
  }

  markProviderTerminal(input: {
    readonly sessionEpoch: number
    readonly responseId: string
    readonly disposition?: Exclude<DeliveryDisposition, 'suppressed'>
  }): boolean {
    const state = this.#byProvider.get(providerKey(input.sessionEpoch, input.responseId))
    if (state === undefined || state.fenced || state.delivered) return false
    state.providerTerminal = true
    state.terminalDisposition = input.disposition ?? 'spoken'
    if (state.nextSequence === 0) this.#retire(state)
    return true
  }

  fenceCurrent(options: {readonly alert?: boolean} = {}): PlaybackGeneration | null {
    const current = this.current
    if (current === null) {
      if (options.alert === true) this.#onAlert(null, null)
      return null
    }
    const fenced = options.alert === true
      ? this.alertFenceGeneration(current)
      : this.switchGeneration(current)
    return fenced ? current : null
  }

  switchGeneration(generation: PlaybackGeneration): boolean {
    return this.#fenceGeneration(generation, false)
  }

  alertFenceGeneration(generation: PlaybackGeneration): boolean {
    return this.#fenceGeneration(generation, true)
  }

  retireClearUnknown(generation: PlaybackGeneration): boolean {
    const state = this.#stateForGeneration(generation)
    if (state === undefined || state.delivered || !state.fenced) return false
    this.#retire(state)
    return true
  }

  markStarted(utteranceId: string, generationEpoch: number): boolean {
    const state = this.#find(utteranceId, generationEpoch)
    if (state === undefined || state.delivered || state.started) return false
    state.started = true
    return true
  }

  recordCleared(
    utteranceId: string,
    generationEpoch: number,
    playedMs: number | null,
  ): PlaybackCompletion | null {
    const state = this.#find(utteranceId, generationEpoch)
    if (state === undefined || state.delivered || !state.fenced) return null
    state.delivered = true
    const completion = this.#completion(
      state,
      state.started || (playedMs !== null && playedMs > 0) ? 'interrupted' : 'suppressed',
      playedMs,
    )
    this.#retire(state)
    return completion
  }

  ackDone(
    utteranceId: string,
    generationEpoch: number,
    playedMs: number | null = null,
  ): PlaybackCompletion | null {
    const state = this.#find(utteranceId, generationEpoch)
    if (
      state === undefined
      || state.fenced
      || state.delivered
      || !state.providerTerminal
      || state.nextSequence === 0
    ) return null
    state.delivered = true
    let disposition: DeliveryDisposition = state.terminalDisposition
    if (
      disposition === 'interrupted'
      && !state.started
      && (playedMs === null || playedMs <= 0)
    ) disposition = 'suppressed'
    const completion = this.#completion(state, disposition, playedMs)
    this.#retire(state)
    return completion
  }

  #completion(
    state: GenerationState,
    disposition: DeliveryDisposition,
    playedMs: number | null,
  ): PlaybackCompletion {
    return {
      session_epoch: state.identity.session_epoch,
      response_id: state.identity.response_id,
      utterance_id: state.identity.utterance_id,
      generation_epoch: state.identity.generation_epoch,
      text: state.transcript,
      disposition,
      started: state.started,
      played_ms: playedMs,
    }
  }

  #find(utteranceId: string, generationEpoch: number): GenerationState | undefined {
    const identity = rendererKey(utteranceId, generationEpoch)
    if (this.#rendererTombstoneSet.has(identity)) return undefined
    return this.#byRenderer.get(identity)
  }

  #fenceGeneration(generation: PlaybackGeneration, alert: boolean): boolean {
    const state = this.#stateForGeneration(generation)
    if (
      state === undefined
      || state.fenced
      || state.delivered
      || this.#currentProviderIdentity !== providerKey(generation.session_epoch, generation.response_id)
    ) return false
    state.fenced = true
    this.#currentProviderIdentity = null
    if (alert) {
      this.#onAlert(generation.utterance_id, generation.generation_epoch)
    } else {
      this.#onClear(generation.utterance_id, generation.generation_epoch)
    }
    return true
  }

  #stateForGeneration(generation: PlaybackGeneration): GenerationState | undefined {
    const state = this.#byRenderer.get(rendererKey(generation.utterance_id, generation.generation_epoch))
    return state !== undefined && sameGeneration(state.identity, generation) ? state : undefined
  }

  #retire(state: GenerationState): void {
    const providerIdentity = providerKey(state.identity.session_epoch, state.identity.response_id)
    const rendererIdentity = rendererKey(state.identity.utterance_id, state.identity.generation_epoch)
    if (this.#byProvider.get(providerIdentity) === state) this.#byProvider.delete(providerIdentity)
    if (this.#byRenderer.get(rendererIdentity) === state) this.#byRenderer.delete(rendererIdentity)
    if (this.#currentProviderIdentity === providerIdentity) this.#currentProviderIdentity = null
    this.#addRendererTombstone(rendererIdentity)
  }

  #addRendererTombstone(rendererIdentity: string): void {
    if (this.#rendererTombstoneSet.has(rendererIdentity)) return
    if (this.#rendererTombstones.length === MAX_RENDERER_TOMBSTONES) {
      const expired = this.#rendererTombstones.shift()
      if (expired !== undefined) this.#rendererTombstoneSet.delete(expired)
    }
    this.#rendererTombstones.push(rendererIdentity)
    this.#rendererTombstoneSet.add(rendererIdentity)
  }

  #requiredId(): string {
    const value = this.#idFactory()
    if (value.length === 0) throw new Error('playback ids must be non-empty')
    return value
  }
}

function providerKey(sessionEpoch: number, responseId: string): string {
  return JSON.stringify([sessionEpoch, responseId])
}

function rendererKey(utteranceId: string, generationEpoch: number): string {
  return JSON.stringify([utteranceId, generationEpoch])
}

function sameGeneration(left: PlaybackGeneration, right: PlaybackGeneration): boolean {
  return left.session_epoch === right.session_epoch
    && left.generation_epoch === right.generation_epoch
    && left.generation_id === right.generation_id
    && left.utterance_id === right.utterance_id
    && left.response_id === right.response_id
}
